lib/AST/Decl.cpp

```diff
@@ -869,6 +869,10 @@ bool Decl::isAlwaysWeakImported() const {
   if (getAttrs().hasAttribute<WeakLinkedAttr>())
     return true;
 
+  if (getAttrs().isUnavailable(getASTContext()))
```

Comment:
I think it would  be better to skip the unavailable case when emitting enum reflection metadata, what do you think?

Reply:
It should be sufficient to check for unavailability in FieldTypeMetadataBuilder.addField(const EnumDecl *enumDecl, const EnumElementDecl *decl, bool hasPayload)...

Reply:
We treat unavailable fields as physically present in the sense that 
```
struct C {
  var x: Int = 0
  @available(macOS, unavailable)
  var y: Int = 0
}

MemoryLayout<C>.size == 16 // true
```

And unavailable functions are there in the object, they are just not available.
```
@available(macOS, unavailable)
public func myUnavailable() {
    print("unavail)")
}

// swiftc -emit-ir < 
define swiftcc void @"$s1A13myUnavailableyyF"() #0 {
```

I felt weak linking is more consistent with the approach that we take elsewhere: it is there but not available

```
internal enum Payload {
    case full(Float)

    @available(macOS, unavailable)
    case half(Float16)
}
```

How many cases does that enum have on macOS one or two (one of which is unavailable)? Today the reflection metadata states that there are two cases. If we change the reflection metadata to not contain the unavailable case we would answer one. Might this break anything?

Reply:
In that case since Float16 participates in the layout, we should be emitting its metadata. Just weak linking it won't work because you'll end up with a null pointer in the reflection metadata and the reflection library won't be able to compute the layout.

What happens if the unavailable type is not POD? Would we try to reference its metadata from the enum's value witness functions for copying, destroying, etc and also fail to link or crash at runtime?